File: Day-01 [22-04-2026]/busly-frontend/src/app/core/services/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { AuditLogDto, BusDto, ConvenienceFeeConfig, CreateRouteRequest, MonthlyRevenueDto, OperatorDto, OperatorRevenueDto, RouteDto, TcVersionDto, UpdateConvenienceFeeRequest } from '../../shared/models/admin.models';

@Injectable({
  providedIn: 'root'
})
export class AdminService {
  private readonly apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  // ─── Routes ────────────────────────────────────────────────────────────────

  /** Create a new route. POST /admin/routes */
  createRoute(request: CreateRouteRequest): Observable<RouteDto> {
    return this.http.post<RouteDto>(`${this.apiUrl}/admin/routes`, request, {
      withCredentials: true
    });
  }

  /** Get all active routes (public). GET /routes */
  getActiveRoutes(): Observable<RouteDto[]> {
    return this.http.get<RouteDto[]>(`${this.apiUrl}/routes`, {
      withCredentials: true
    });
  }

  /**
   * Get all routes (including inactive) for admin management. GET /admin/routes
   */
  getAllRoutes(): Observable<RouteDto[]> {
    return this.http.get<RouteDto[]>(`${this.apiUrl}/admin/routes`, {
      withCredentials: true
    });
  }

  /** Toggle a route's active status. PATCH /admin/routes/:id/toggle */
  toggleRoute(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/routes/${id}/toggle`, {}, {
      withCredentials: true
    });
  }

  // ─── Operators ─────────────────────────────────────────────────────────────

  /** Get all operators pending approval. GET /admin/operators/pending */
  getPendingOperators(): Observable<OperatorDto[]> {
    return this.http.get<OperatorDto[]>(`${this.apiUrl}/admin/operators/pending`, {
      withCredentials: true
    });
  }

  /** Get all operators for management. GET /admin/operators */
  getAllOperators(): Observable<OperatorDto[]> {
    return this.http.get<OperatorDto[]>(`${this.apiUrl}/admin/operators`, {
      withCredentials: true
    });
  }

  /** Approve a pending operator. PATCH /admin/operators/:id/approve */
  approveOperator(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/operators/${id}/approve`, {}, {
      withCredentials: true
    });
  }

  /** Reject a pending operator. PATCH /admin/operators/:id/reject */
  rejectOperator(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/operators/${id}/reject`, {}, {
      withCredentials: true
    });
  }

  /** Toggle an operator's active/suspended status. PATCH /admin/operators/:id/toggle */
  toggleOperator(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/operators/${id}/toggle`, {}, {
      withCredentials: true
    });
  }

  // ─── Buses ─────────────────────────────────────────────────────────────────

  /** Get all buses pending approval. GET /admin/buses/pending */
  getPendingBuses(): Observable<BusDto[]> {
    return this.http.get<BusDto[]>(`${this.apiUrl}/admin/buses/pending`, {
      withCredentials: true
    });
  }

  /** Get all buses for management. GET /admin/buses */
  getAllBuses(): Observable<BusDto[]> {
    return this.http.get<BusDto[]>(`${this.apiUrl}/admin/buses`, {
      withCredentials: true
    });
  }

  /** Get all buses by specific operator. GET /admin/operators/:id/buses */
  getBusesByOperator(operatorId: string): Observable<BusDto[]> {
    return this.http.get<BusDto[]>(`${this.apiUrl}/admin/operators/${operatorId}/buses`, {
      withCredentials: true
    });
  }

  /** Approve a pending bus. PATCH /admin/buses/:id/approve */
  approveBus(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/buses/${id}/approve`, {}, {
      withCredentials: true
    });
  }

  /** Reject a pending bus. PATCH /admin/buses/:id/reject */
  rejectBus(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/buses/${id}/reject`, {}, {
      withCredentials: true
    });
  }

  /** Toggle a bus's operational status. PATCH /admin/buses/:id/toggle */
  toggleBusStatus(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/buses/${id}/toggle`, {}, {
      withCredentials: true
    });
  }

  // ─── Revenue ───────────────────────────────────────────────────────────────

  /** Get monthly platform revenue. GET /admin/revenue */
  getRevenue(): Observable<MonthlyRevenueDto[]> {
    return this.http.get<MonthlyRevenueDto[]>(`${this.apiUrl}/admin/revenue`, {
      withCredentials: true
    });
  }

  /** Get revenue broken down by operator. GET /admin/revenue/by-operator */
  getRevenueByOperator(): Observable<OperatorRevenueDto[]> {
    return this.http.get<OperatorRevenueDto[]>(`${this.apiUrl}/admin/revenue/by-operator`, {
      withCredentials: true
    });
  }

  /** Get system audit logs. GET /admin/audit-logs */
  getAuditLogs(): Observable<AuditLogDto[]> {
    return this.http.get<AuditLogDto[]>(`${this.apiUrl}/admin/audit-logs`, {
      withCredentials: true
    });
  }

  /** Get all T&C versions. GET /admin/tc */
  getAllTcVersions(): Observable<TcVersionDto[]> {
    return this.http.get<TcVersionDto[]>(`${this.apiUrl}/admin/tc`, {
      withCredentials: true
    });
  }

  /** Publish new T&C version. POST /admin/tc */
  publishTc(version: string, content: string): Observable<any> {
    return this.http.post(`${this.apiUrl}/admin/tc`, { version, content }, {
      withCredentials: true
    });
  }

  // ─── Platform Config ───────────────────────────────────────────────────────

  /** Get current convenience fee config. GET /admin/config/convenience-fee */
  getConvenienceFeeConfig(): Observable<ConvenienceFeeConfig> {
    return this.http.get<ConvenienceFeeConfig>(`${this.apiUrl}/admin/config/convenience-fee`, {
      withCredentials: true
    });
  }

  /** Update convenience fee config. PATCH /admin/config/convenience-fee */
  updateConvenienceFeeConfig(request: UpdateConvenienceFeeRequest): Observable<any> {
    return this.http.patch(`${this.apiUrl}/admin/config/convenience-fee`, request, {
      withCredentials: true
    });
  }
}

File: Day-01 [22-04-2026]/Busly.API/Services/AdminService.cs
using Busly.API.DTOs.Admin;
using Busly.API.Repositories;
using BuslyRoute = Busly.API.Models.Route;
using BusOperator = Busly.API.Models.BusOperator;
using Bus = Busly.API.Models.Bus;
using TcVersion = Busly.API.Models.TcVersion;

namespace Busly.API.Services;

public class AdminService : IAdminService
{
    private readonly IRouteRepository _routeRepo;
    private readonly IAuditRepository _auditRepo;
    private readonly IAuthRepository _authRepo;
    private readonly IBookingRepository _bookingRepo;
    private readonly ITcRepository _tcRepo;

    public AdminService(
        IRouteRepository routeRepo,
        IAuditRepository auditRepo,
        IAuthRepository authRepo,
        IBookingRepository bookingRepo,
        ITcRepository tcRepo)
    {
        _routeRepo   = routeRepo;
        _auditRepo   = auditRepo;
        _authRepo    = authRepo;
        _bookingRepo = bookingRepo;
        _tcRepo      = tcRepo;
    }

    // ── Routes ────────────────────────────────────────────────────────────────

    public async Task<RouteDto> CreateRouteAsync(CreateRouteRequest request, Guid adminId)
    {
        var existingRoute = await _routeRepo.GetByCitiesAsync(request.SourceCity, request.DestinationCity);
        
        if (existingRoute != null)
        {
            if (existingRoute.IsActive)
            {
                throw new InvalidOperationException("Route already exists and is active.");
            }

            // Re-activate the existing inactive route
            existingRoute.IsActive = true;
            await _routeRepo.UpdateAsync(existingRoute);
            await _auditRepo.LogAsync(adminId, "admin", "REACTIVATE_ROUTE", "route", existingRoute.Id);
            // Return 200-style DTO with a flag so the caller knows it was a reactivation, not a new create
            return MapRoute(existingRoute);
        }

        var route = new BuslyRoute
        {
            Id             = Guid.NewGuid(),
            SourceCity     = request.SourceCity,
            DestinationCity = request.DestinationCity,
            IsActive       = true,
            CreatedByAdmin = adminId
        };

        await _routeRepo.CreateAsync(route);
        await _auditRepo.LogAsync(adminId, "admin", "CREATE_ROUTE", "route", route.Id, 
            $"{{\"source\":\"{route.SourceCity}\", \"destination\":\"{route.DestinationCity}\"}}");

        return MapRoute(route);
    }

    public async Task<List<RouteDto>> GetAllRoutesAsync()
    {
        var routes = await _routeRepo.GetAllAsync();
        return routes.Select(MapRoute).ToList();
    }

    public async Task<List<RouteDto>> GetActiveRoutesAsync()
    {
        var routes = await _routeRepo.GetAllActiveAsync();
        return routes.Select(MapRoute).ToList();
    }

    public async Task ToggleRouteAsync(Guid routeId, Guid adminId)
    {
        await _routeRepo.ToggleAsync(routeId);
        await _auditRepo.LogAsync(adminId, "admin", "TOGGLE_ROUTE", "route", routeId);
    }

    // ── Operators ─────────────────────────────────────────────────────────────

    public async Task<List<OperatorDto>> GetPendingOperatorsAsync()
    {
        var operators = await _authRepo.GetPendingOperatorsAsync();
        return operators.Select(MapOperator).ToList();
    }

    public async Task<List<OperatorDto>> GetAllOperatorsAsync()
    {
        var operators = await _authRepo.GetAllOperatorsAsync();
        return operators.Select(MapOperator).ToList();
    }

    public async Task ApproveOperatorAsync(Guid operatorId, Guid adminId)
    {
        await _authRepo.UpdateOperatorStatusAsync(operatorId, "APPROVED", adminId);
        await _auditRepo.LogAsync(adminId, "admin", "APPROVE_OPERATOR", "bus_operator", operatorId, "{\"status\":\"APPROVED\"}");
    }

    public async Task RejectOperatorAsync(Guid operatorId, Guid adminId)
    {
        await _authRepo.UpdateOperatorStatusAsync(operatorId, "REJECTED");
        await _auditRepo.LogAsync(adminId, "admin", "REJECT_OPERATOR", "bus_operator", operatorId);
    }

    public async Task ToggleOperatorAsync(Guid operatorId, Guid adminId)
    {
        var op = await _authRepo.GetOperatorByIdAsync(operatorId)
            ?? throw new InvalidOperationException("Operator not found");

        var newStatus = op.Status == "APPROVED" ? "DISABLED" : "APPROVED";

        await _authRepo.UpdateOperatorStatusAsync(operatorId, newStatus);
        await _auditRepo.LogAsync(adminId, "admin", "TOGGLE_OPERATOR", "bus_operator", operatorId);
    }

    // ── Buses ─────────────────────────────────────────────────────────────────

    public async Task<List<BusDto>> GetPendingBusesAsync()
    {
        var buses = await _authRepo.GetPendingBusesAsync();
        return buses.Select(MapBus).ToList();
    }

    public async Task<List<BusDto>> GetAllBusesAsync()
    {
        var buses = await _authRepo.GetAllBusesAsync();
        return buses.Select(MapBus).ToList();
    }

    public async Task<List<BusDto>> GetBusesByOperatorAsync(Guid operatorId)
    {
        var buses = await _authRepo.GetBusesByOperatorAsync(operatorId);
        return buses.Select(MapBus).ToList();
    }

    public async Task ApproveBusAsync(Guid busId, Guid adminId)
    {
        await _authRepo.UpdateBusStatusAsync(busId, "ACTIVE", adminId);
        await _auditRepo.LogAsync(adminId, "admin", "APPROVE_BUS", "bus", busId, "{\"status\":\"ACTIVE\"}");
    }

    public async Task RejectBusAsync(Guid busId, Guid adminId)
    {
        await _authRepo.UpdateBusStatusAsync(busId, "REJECTED", adminId);
        await _auditRepo.LogAsync(adminId, "admin", "REJECT_BUS", "bus", busId);
    }

    public async Task ToggleBusStatusAsync(Guid busId, Guid adminId)
    {
        var bus = await _authRepo.GetBusByIdAsync(busId);
        if (bus == null) throw new InvalidOperationException("Bus not found");

        if (bus.Status == "PENDING")
            throw new InvalidOperationException("Cannot toggle a pending bus. Use Approve/Reject.");

        var newStatus = bus.Status == "ACTIVE" ? "DISABLED" : "ACTIVE";
        await _authRepo.UpdateBusStatusAsync(busId, newStatus, adminId);
        await _auditRepo.LogAsync(adminId, "admin", "TOGGLE_BUS_STATUS", "bus", busId, $"{{\"status\":\"{newStatus}\"}}");
    }

    // ── Revenue ───────────────────────────────────────────────────────────────

    public async Task<List<MonthlyRevenueDto>> GetMonthlyRevenueAsync()
    {
        return await _bookingRepo.GetMonthlyRevenueAsync();
    }

    public async Task<List<OperatorRevenueDto>> GetOperatorRevenueAsync()
    {
        return await _bookingRepo.GetOperatorRevenueAsync();
    }

    // ── T&C Management ────────────────────────────────────────────────────────

    public async Task PublishTcVersionAsync(CreateTcRequest request, Guid adminId)
    {
        var tc = new TcVersion
        {
            Id               = Guid.NewGuid(),
            Version          = request.Version,
            Content          = request.Content,
            PublishedAt      = DateTime.UtcNow,
            EffectiveAt      = request.EffectiveAt ?? DateTime.UtcNow,
            PublishedByAdmin = adminId,
            IsActive         = true
        };

        await _tcRepo.CreateAsync(tc);
        await _tcRepo.DeactivateAllExceptAsync(tc.Id);
        
        await _auditRepo.LogAsync(adminId, "admin", "PUBLISH_TC", "tc_version", tc.Id, $"{{\"version\":\"{tc.Version}\"}}");
    }

    public async Task<List<TcVersionDto>> GetAllTcVersionsAsync()
    {
        var versions = await _tcRepo.GetAllAsync();
        return versions.Select(v => new TcVersionDto
        {
            Id          = v.Id,
            Version     = v.Version,
            Content     = v.Content,
            PublishedAt = v.PublishedAt,
            EffectiveAt = v.EffectiveAt,
            IsActive    = v.IsActive
        }).ToList();
    }

    public async Task<TcVersionDto?> GetCurrentTcAsync()
    {
        var activeTc = await _tcRepo.GetActiveAsync();
        
        if (activeTc == null)
            return null;

        return new TcVersionDto
        {
            Id = activeTc.Id,
            Version = activeTc.Version,
            Content = activeTc.Content,
            PublishedAt = activeTc.PublishedAt,
            EffectiveAt = activeTc.EffectiveAt,
            IsActive = activeTc.IsActive
        };
    }

    public async Task<List<AuditLogDto>> GetAuditLogsAsync()
    {
        var logs = await _auditRepo.GetAllAsync();
        return logs.Select(MapAuditLog).ToList();
    }

    // ── Mappers ───────────────────────────────────────────────────────────────

    private static RouteDto MapRoute(BuslyRoute r) => new()
    {
        Id              = r.Id,
        SourceCity      = r.SourceCity,
        DestinationCity = r.DestinationCity,
        IsActive        = r.IsActive
    };

    private static OperatorDto MapOperator(BusOperator op) => new()
    {
        Id          = op.Id,
        CompanyName = op.CompanyName,
        Email       = op.Email,
        Phone       = op.Phone,
        ContactEmail = op.Email,
        ContactPhone = op.Phone,
        Status      = op.Status,
        CreatedAt   = op.CreatedAt,
        TcVersion   = op.TcVersion,
        TcAcceptedAt = op.TcAcceptedAt
    };

    private static BusDto MapBus(Bus b) => new()
    {
        Id              = b.Id,
        BusNumber       = b.BusNumber,
        BusName         = b.BusName,
        OwnerName       = b.OwnerName,
        OwnerPhone      = b.OwnerPhone,
        OwnerEmail      = b.OwnerEmail,
        BasePrice       = b.BasePrice,
        DriverName      = b.DriverName,
        DriverPhone     = b.DriverPhone,
        ConductorName   = b.ConductorName,
        ConductorPhone  = b.ConductorPhone,
        SourceCity      = b.Route?.SourceCity,
        DestinationCity = b.Route?.DestinationCity,
        LayoutName      = b.Layout?.LayoutName,
        Status          = b.Status,
        OperatorId      = b.OperatorId,
        RouteId         = b.RouteId,
        CreatedAt       = b.CreatedAt,
        OperatingDays   = b.OperatingDays?.Select(od => new OperatingDayDto 
        { 
            DayOfWeek = od.DayOfWeek, 
            IsActive = od.IsActive 
        }).ToList() ?? new List<OperatingDayDto>()
    };

    private static AuditLogDto MapAuditLog(Busly.API.Models.AuditLog log) => new()
    {
        Id         = log.Id,
        ActorId    = log.ActorId,
        ActorRole  = log.ActorRole ?? "unknown",
        Action     = log.Action,
        EntityType = log.EntityType,
        EntityId   = log.EntityId,
        Metadata   = log.Metadata,
        Timestamp  = log.PerformedAt
    };
}
